File: src/commands/git/prune.ts
import * as prompts from "@clack/prompts";
import { Command } from "commander";
import { getShellError } from "../../utils/errors";

export function parseBranchList(raw: string): {
	currentBranch: string;
	branches: string[];
} {
	const lines = raw.split("\n").filter((l) => l.trim());
	let currentBranch = "";
	const branches: string[] = [];

	for (const line of lines) {
		const name = line.replace("*", "").trim();
		if (line.trimStart().startsWith("*")) {
			currentBranch = name;
		} else {
			branches.push(name);
		}
	}

	return { currentBranch, branches };
}

export const prune = new Command("prune")
	.description("Interactively delete local git branches")
	.argument("[pattern]", "Filter branches by substring")
	.option("-f, --force", "Skip confirmation prompt")
	.action(async (pattern: string | undefined, opts: { force?: boolean }) => {
		const result = await Bun.$`git branch`.text();
		const { currentBranch, branches } = parseBranchList(result);

		const filtered = pattern
			? branches.filter((b) => b.includes(pattern))
			: branches;

		if (filtered.length === 0) {
			console.log(
				pattern
					? `No branches matching "${pattern}" (excluding current: ${currentBranch})`
					: `No branches to prune (only current: ${currentBranch})`,
			);
			return;
		}

		let toDelete: string[];

		if (opts.force) {
			toDelete = filtered;
		} else {
			prompts.intro("git branch prune");

			const selected = await prompts.multiselect({
				message: `Select branches to delete (current: ${currentBranch})`,
				options: filtered.map((b) => ({ value: b, label: b })),
				required: true,
			});

			if (prompts.isCancel(selected)) {
				prompts.cancel("Cancelled.");
				process.exit(0);
			}

			toDelete = selected as string[];
		}

		if (toDelete.length === 0) {
			console.log("No branches selected.");
			return;
		}

		for (const branch of toDelete) {
			try {
				await Bun.$`git branch -D ${branch}`.quiet();
				console.log(`  Deleted ${branch}`);
			} catch (e: unknown) {
				console.error(`  Failed to delete ${branch}: ${getShellError(e)}`);
			}
		}

		console.log(
			`\nPruned ${toDelete.length} branch${toDelete.length > 1 ? "es" : ""}.`,
		);
	});

File: src/commands/git/prune.rs
use anyhow::Result;
use dialoguer::{MultiSelect, theme::ColorfulTheme};

pub fn parse_branch_list(raw: &str) -> (String, Vec<String>) {
    let mut current = String::new();
    let mut branches = Vec::new();

    for line in raw.lines() {
        let name = line.replace('*', "").trim().to_string();
        if name.is_empty() { continue; }
        if line.trim_start().starts_with('*') {
            current = name;
        } else {
            branches.push(name);
        }
    }

    (current, branches)
}

fn run_cmd(args: &[&str]) -> Result<String> {
    let output = std::process::Command::new("git").args(args).output()?;
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

pub fn execute_prune(pattern: Option<&str>, force: bool) -> Result<()> {
    let raw = run_cmd(&["branch"])?;
    let (current, branches) = parse_branch_list(&raw);

    let filtered: Vec<String> = if let Some(pat) = pattern {
        branches.into_iter().filter(|b| b.contains(pat)).collect()
    } else {
        branches
    };

    if filtered.is_empty() {
        match pattern {
            Some(pat) => println!("No branches matching \"{pat}\" (excluding current: {current})"),
            None => println!("No branches to prune (only current: {current})"),
        }
        return Ok(());
    }

    let to_delete: Vec<String> = if force {
        filtered.clone()
    } else {
        let selections = MultiSelect::with_theme(&ColorfulTheme::default())
            .with_prompt(format!("Select branches to delete (current: {current})"))
            .items(&filtered.iter().map(|s| s.as_str()).collect::<Vec<_>>())
            .interact()?;
        selections.into_iter().map(|i| filtered[i].clone()).collect()
    };

    if to_delete.is_empty() {
        println!("No branches selected.");
        return Ok(());
    }

    for branch in &to_delete {
        let status = std::process::Command::new("git")
            .args(["branch", "-D", branch])
            .status()?;
        if status.success() {
            println!("  Deleted {branch}");
        } else {
            eprintln!("  Failed to delete {branch}");
        }
    }

    let count = to_delete.len();
    let plural = if count == 1 { "" } else { "es" };
    println!("\nPruned {count} branch{plural}.");

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_branch_list_identifies_current() {
        let raw = "  main\n* feature-x\n";
        let (current, branches) = parse_branch_list(raw);
        assert_eq!(current, "feature-x");
        assert_eq!(branches, vec!["main"]);
    }

    #[test]
    fn test_parse_branch_list_multiple() {
        let raw = "  main\n  develop\n* feature-x\n";
        let (current, branches) = parse_branch_list(raw);
        assert_eq!(current, "feature-x");
        assert_eq!(branches, vec!["main", "develop"]);
    }

    #[test]
    fn test_parse_branch_list_with_slashes() {
        let raw = "  main\n  feature/foo\n* feature/bar\n";
        let (current, branches) = parse_branch_list(raw);
        assert_eq!(current, "feature/bar");
        assert_eq!(branches, vec!["main", "feature/foo"]);
    }

    #[test]
    fn test_parse_branch_list_whitespace_handling() {
        let raw = "  main\n  \n* feature-x\n";
        let (_current, branches) = parse_branch_list(raw);
        assert_eq!(branches, vec!["main"]);
    }

    #[test]
    fn test_parse_branch_list_empty() {
        let (current, branches) = parse_branch_list("");
        assert!(current.is_empty());
        assert!(branches.is_empty());
    }
}
